### PR TITLE
feat: AaveCLRobotOperator contract for admin actions

### PR DESCRIPTION
**Pull request by brotherlymite**

### CODE REVIEW

**Comment by eboadom** [comment]
I think the `AaveCLRobotOperator` should also have the function to refill the keepers, even if permissionless. To gather all "admin" actions (refilling kind of is one) in the same entry point contract

**Comment by eboadom** [comment]
Any reason @brotherlymite to not allow changing the admins?

**Comment by eboadom** [comment]
Some events are missing I think

File: src/contracts/AaveCLRobotOperator.sol [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      0x0,
+      upkeepContract,
+      gasLimit,
+      address(this),
+      checkData,
+      amountToFund,
+      0,
+      address(this)
+    );
+    bytes4 registerSig = IKeeperRegistrar.register.selector;
```
Comment (eboadom) [inline src/contracts/AaveCLRobotOperator.sol:90]: This local var doesn't seem needed. Just used once afterwards

File: src/contracts/AaveCLRobotOperator.sol [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    uint32 gasLimit,
+    bytes memory checkData,
+    uint96 amountToFund,
+    address keeperRegistry,
+    address keeperRegistrar
+  ) external onlyFundsAdmin returns (uint256) {
+    (IKeeperRegistry.State memory state, , ) = IKeeperRegistry(keeperRegistry).getState();
+    uint256 oldNonce = state.nonce;
+
+    bytes memory payload = abi.encode(
```
Comment (eboadom) [inline src/contracts/AaveCLRobotOperator.sol:79]: Better to document a bit the params passed, with comment on each line

File: src/contracts/AaveCLRobotOperator.sol [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    );
+    bytes4 registerSig = IKeeperRegistrar.register.selector;
+    LinkTokenInterface(LINK_TOKEN).transferAndCall(
+      keeperRegistrar,
+      amountToFund,
+      bytes.concat(registerSig, payload)
+    );
+
+    (state, , ) = IKeeperRegistry(keeperRegistry).getState();
+    if (state.nonce == oldNonce + 1) {
```
Comment (eboadom) [inline src/contracts/AaveCLRobotOperator.sol:98]: What exactly is this condition for? I assume the auto-approve, but better to document a bit what is it about

**Comment by brotherlymite** [comment]
> Any reason @brotherlymite to not allow changing the admins?

Currently on the chainlink contract there is no function to change the admin, so the only way is to cancel, withdrawFunds and register a new keeper.

**Comment by eboadom** [comment]
> > Currently on the chainlink contract there is no function to change the admin, so the only way is to cancel, withdrawFunds and register a new keeper.

Sure, but I'm talking to the operator. The admin on the keepers will be the operator, and precisely the admins on the operator are to abstract that. Practical case: migration governance v2 to v3

**Comment by brotherlymite** [comment]
> I think the `AaveCLRobotOperator` should also have the function to refill the keepers, even if permissionless. To gather all "admin" actions (refilling kind of is one) in the same entry point contract

The issue is for that we need to send LINK to the `AaveCLRobotOperator` contract and then the contract calls the `addFunds()` on the chainlink registry which uses `transferFrom(msg.sender, chainlinkRegistry)` so we spend some extra gas on an additional transfer to the `AaveCLRobotOperator` contract.

**Comment by brotherlymite** [comment]
> `withdrawLink()` should not be permissionless

`withdrawLink()` only withdraws LINK to the withdraw address set. After cancelling the keeper we need to wait for some blocks to withdraw LINK so its quite tricky to use the executor to withdraw link after cancelling the keeper, the other option is to use the maintenance admin to withdraw link to the set withdraw address

**Comment by brotherlymite** [comment]
> > > Currently on the chainlink contract there is no function to change the admin, so the only way is to cancel, withdrawFunds and register a new keeper.
> 
> Sure, but I'm talking about the operator. The admin on the keepers will be the operator, and precisely the admins on the operator are to abstract that. Practical case: migration governance v2 to v3

Ah okay got confused, have added. For migration of existing keepers there is no way to change the admin so for that we need to cancel them and register the new ones with the operator contract.

File: src/contracts/AaveCLRobotOperator.sol (hunk 3 above)
Comment (brotherlymite) [reply src/contracts/AaveCLRobotOperator.sol:100]: Yes the nonce increments when keeper has been registered and is auto-approved, have added some comments

File: src/contracts/AaveCLRobotOperator.sol (hunk 2 above)
Comment (brotherlymite) [reply src/contracts/AaveCLRobotOperator.sol:80]: Fixed

**Comment by eboadom** [comment]
> The issue is for that we need to send LINK to the `AaveCLRobotOperator` contract and then the contract calls the `addFunds()` on the chainlink registry which uses `transferFrom(msg.sender, chainlinkRegistry)` so we spend some extra gas on an additional transfer to the `AaveCLRobotOperator` contract.

Remember what we discussed. Gas is not a priority when the action is not frequent, architecture and UX is

**Review by eboadom — APPROVED** [review]

**Pull request merged**